gratipay/models/participant/email.py

```diff
@@ -225,22 +226,51 @@ def finish_email_verification(self, email, nonce):
           updated if applicable (``None`` if not).
 
         """
+        _fail = VERIFICATION_FAILED, None, None
         if '' in (email.strip(), nonce.strip()):
-            return VERIFICATION_FAILED, None, None
+            return _fail
         with self.db.get_cursor() as cursor:
+
+            # Load an email record. Check for an address match, but don't check
+            # the nonce at this point. We want to compare in constant time to
+            # avoid timing attacks, and we'll do that below.
+
             record = self.get_email(email, cursor, and_lock=True)
             if record is None:
-                return VERIFICATION_FAILED, None, None
-            packages = self.get_packages_claiming(cursor, nonce)
-            if record.verified and not packages:
-                assert record.nonce is None  # and therefore, order of conditions matters
+
+                # We don't have that email address on file. Maybe it used to be
+                # on file but was explicitly removed (they followed an old link
+                # after removing in the UI?), or maybe it was never on file in
+                # the first place (they munged the querystring?).
+
+                return _fail
+
+            if record.nonce is None:
```

Comment:
Why should we rely on the `nonce` being empty to deduce that the address was marked as verified, when we could just check `record.verified`? 🤔